### PR TITLE
Simplify forward pass in LoRALinear layers

### PR DESCRIPTION
**Pull request by Andrei-Aksionov**
Hi there 👋 

This PR contains simplification of the logic in `forward` methods.
Now there is no need for so many if statements.
Essentially we need to do the forward pass with the pretrained weights *anyway* and add to the result output from LoRA weights if the conditions are met. Thus only one if statement is needed.

In addition in this PR I also removed transposing function from each merge/forward method and created a class method instead.

### CODE REVIEW

File: lit_gpt/lora.py [hunk 1]
```diff
@@ -402,9 +386,7 @@ def T(w):
                 after_A.transpose(-2, -1),  # (64, 64, 4) -> (64, 4, 64)
                 self.lora_B.unsqueeze(-1),  # (256, 2) -> (256, 2, 1)
                 groups=sum(self.enable_lora),
-            ).transpose(
-                -2, -1
-            )  # (64, 4, 64) @ (256, 2, 1) -> (64, 256, 64) -> (64, 64, 256)
+            ).transpose(-2, -1)  # (64, 4, 64) @ (256, 2, 1) -> (64, 256, 64) -> (64, 64, 256)
```
Comment (Andrei-Aksionov) [inline lit_gpt/lora.py:389]: I also have a question regarding `.transpose(-2, -1)`. 
It has a shorter equivalent - [`.mT`](https://pytorch.org/docs/stable/tensors.html#torch.Tensor.mT), but since `.transpose` is used everywhere it looks like it's better to stick to it.
Any thought on it?

File: lit_gpt/lora.py [hunk 2]
```diff
@@ -373,25 +366,16 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
             Output tensor of shape (batch_size, context_length, 3 * embedding_size)
         """
 
-        def T(w):
-            return w.T if self.fan_in_fan_out else w
```
Comment (Andrei-Aksionov) [inline lit_gpt/lora.py:377]: In `LoRALinear` for transposing `w.transpose(0,1)` is called, while for `LoRAQKVLinear` - `w.T`. 
So using the same transposing function for both classes might look like a wrong decision, because with `.T` method we are reversing the whole shape of a tensor.
But here we apply transposing function to only weights that are two dimensional, so it's all good. Annotations with shapes really help here.

If you feel uncomfortable about it I can override the function for `LoRAQKVLinear` like this:
```python
def T(self, x: torch.Tensor) -> torch.Tensor:
    """Transpose input tensor if weights are stored in format (fan_in, fan_out)"""
    return x.permute(*torch.arange(x.ndim -1, -1, -1)) if self.fan_in_fan_out else x
```
which looks convoluted because:
> UserWarning: The use of `x.T` on tensors of dimension other than 2 to reverse their shape is deprecated and it will throw an error in a future release. Consider `x.mT` to transpose batches of matrices or `x.permute(*torch.arange(x.ndim - 1, -1, -1))` to reverse the dimensions of a tensor.

File: lit_gpt/lora.py [hunk 3]
```diff
@@ -373,25 +366,16 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
-        # and do the summation (as per scheme at the top of the file)
-        if self.merged:
-            return F.linear(x, T(self.weight), bias=self.bias)
-        # `F.linear` automatically transposes the second argument (T(self.weight) in our case)
-        result = F.linear(x, T(self.weight), bias=self.bias)  # (64, 64, 128) @ (384, 128) -> (64, 64, 384)
-        if self.r > 0 and any(self.enable_lora):
+        # if weights are merged or LoRA is disabled (r <= 0 or all `enable_lora` are False) - it's only a regular nn.Linear forward pass;
+        # otherwise in addition do the forward pass with LoRA weights and add it's output to the output from pretrained weights
+        result = F.linear(x, self.T(self.weight), bias=self.bias)
+        if all((self.r > 0, any(self.enable_lora), not self.merged)):
```
Comment (awaelchli) [inline lit_gpt/lora.py:378]:
```suggestion
        if self.r > 0 and any(self.enable_lora) and not self.merged):
```
An argument can be made that this is more readable to the human

**Review by awaelchli — APPROVED** [review]
Looks great. 

Did you also sanity-check that the loss is decreasing by kicking off a lora training? I can run it if you need help.

File: lit_gpt/lora.py (hunk 1 above)
Comment (awaelchli) [reply lit_gpt/lora.py:389]: My personal preference is to keep the transpose, spelling out what it does should help people who are onboarding and learning the operations in a transformer model.

File: lit_gpt/lora.py (hunk 2 above)
Comment (awaelchli) [reply lit_gpt/lora.py:377]: The dimension is 2, so there should be no need to generalize to a form that is harder to read.

Comment (Andrei-Aksionov) [reply lit_gpt/lora.py:377]: Agree.
Just wanted to provide options.

Want to clarify: since the dimension is 2, `.transpose(0, 1)` == `.T`, thus we can reuse in `LoRAQKVLinear` method from `LoRALinear` and you are not against it, right?

File: lit_gpt/lora.py (hunk 3 above)
Comment (Andrei-Aksionov) [reply lit_gpt/lora.py:378]: With `all(...)` you can read it as:
> if `all` conditions are met then ...

With `and`:
> if this is true `and` this is true `and` ...

So I guess it's a personal preference. For me `all(...)` looks cleaner (and I usually prefer it when I need more than one `and` operator). 
But it's your repo, you are the boss, so ... 😄

**Comment by Andrei-Aksionov** [comment]
> Did you also sanity-check that the loss is decreasing by kicking off a lora training? I can run it if you need help.

Now I have an access to a server with a GPU so I can run it.
Just want to clarify: sanity-check includes just verifying that the loss is decreasing or I need to run training for at least `X` iterations?

File: lit_gpt/lora.py (hunk 3 above)
Comment (awaelchli) [reply lit_gpt/lora.py:366]: I'll apply the suggestion to keep consistency with the rest of the code base ([example](https://github.com/Lightning-AI/lit-gpt/blob/d0f082f164a8be5d3696b9e7d326b9d190ab9ac1/lit_gpt/model.py#L260C7-L265))

File: lit_gpt/lora.py (hunk 2 above)
Comment (awaelchli) [reply lit_gpt/lora.py:365]: I would say `.T` and `.transpose` is what the average PyTorch user knows and can interpret without looking up the docs, so let's stick to these.

File: lit_gpt/lora.py [hunk 4]
```diff
@@ -361,25 +354,16 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
-        # and do the summation (as per scheme at the top of the file)
-        if self.merged:
-            return F.linear(x, T(self.weight), bias=self.bias)
-        # `F.linear` automatically transposes the second argument (T(self.weight) in our case)
-        result = F.linear(x, T(self.weight), bias=self.bias)  # (64, 64, 128) @ (384, 128) -> (64, 64, 384)
-        if self.r > 0 and any(self.enable_lora):
+        # if weights are merged or LoRA is disabled (r <= 0 or all `enable_lora` are False) - it's only a regular nn.Linear forward pass;
+        # otherwise in addition do the forward pass with LoRA weights and add it's output to the output from pretrained weights
+        result = F.linear(x, self.T(self.weight), bias=self.bias)
+        if self.r > 0 and any(self.enable_lora) and not self.merged):
```
Comment (awaelchli) [inline lit_gpt/lora.py:366]:
```suggestion
        if self.r > 0 and any(self.enable_lora) and not self.merged:
```

File: lit_gpt/lora.py (hunk 3 above)
Comment (Andrei-Aksionov) [reply lit_gpt/lora.py]: Makes sense.
